src/librustc/traits/error_reporting.rs

```diff
@@ -2112,9 +2112,13 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 err.note(&format!("required by cast to type `{}`",
                                   self.ty_to_string(target)));
             }
-            ObligationCauseCode::RepeatVec => {
+            ObligationCauseCode::RepeatVec(suggest_const_in_array_repeat_expression) => {
                 err.note("the `Copy` trait is required because the \
                           repeated element will be copied");
+                if suggest_const_in_array_repeat_expression {
+                    err.help("add `#![feature(const_in_array_repeat_expression)]` to the crate \
+                              attributes to enable");
```

Comment:
Nit: could we add a line to the message that suggests *why* `const_in_array_repeat_expression` will work for some array initializers but not others? Something like, "this array initializer is eligible for promotion, add `#![feature(const_in_array_repeat_expression)]` to the crate attributes to enable"?

Reply:
"eligible for promotion" sounds a bit jargon-packed -- can we make it more newbie friendly? cc @estebank

Reply:
I don’t mind adding some wording here but I’m not sure if there are any other cases where the “enabling this feature would help” message contains more context. I also don’t have any better suggestions for the exact wording we add. 

It just occurs to me now, do we normally only add these help messages on nightly builds?

Reply:
"is eligible for promotion" => "could be evaluated at compile-time"?

Reply:
On a stable/beta compiler we'd do something like this:
```rust
error[E0658]: or-patterns syntax is experimental
 --> src/lib.rs:3:14
  |
3 |         Some(0 | 1) => {}
  |              ^^^^^
  |
  = note: for more information, see https://github.com/rust-lang/rust/issues/54883
```
and on nightly:
```rust

error[E0658]: or-patterns syntax is experimental
 --> src/lib.rs:3:14
  |
3 |         Some(0 | 1) => {}
  |              ^^^^^
  |
  = note: for more information, see https://github.com/rust-lang/rust/issues/54883
  = help: add `#![feature(or_patterns)]` to the crate attributes to enable
```
(This is automated by `emit_feature_err`.)

Reply:
I've adjusted the error message so that it links to #49147 on stable/beta, and on nightly suggests to add the feature attribute with a short description of why it is applicable.

Reply:
LGTM.

Reply:
I've pushed again to add the short description to the message which shows on stable/beta, not just nightly - otherwise it's quite unclear why the "for more information" note is shown.